components/places/src/history_sync/plan.rs

```diff
@@ -0,0 +1,619 @@
+    apply_synced_deletion,
+    apply_synced_visits,
+    apply_synced_reconcilliation,
+    fetch_outgoing,
+    fetch_visits,
+    FetchedVisit,
+    FetchedVisitPage,
+    finish_outgoing,
+    OutgoingInfo,
+};
+
+use api::history::{can_add_url};
+use error::*;
+
+use sync15_adapter::{
+    IncomingChangeset,
+    OutgoingChangeset,
+    Payload,
+};
+
+// In desktop sync, bookmarks are clamped to Jan 23, 1993 (which is 727747200000)
+// There's no good reason history records could be older than that, so we do
+// the same here (even though desktop's history currently doesn't)
+// XXX - there's probably a case to be made for this being, say, 5 years ago -
+// then all requests earlier than that are collapsed into a single visit at
+// this timestamp.
+const EARLIEST_TIMESTAMP: Timestamp = Timestamp(727747200000);
+
+/// Clamps a history visit date between the current date and the earliest
+/// sensible date.
+fn clamp_visit_date(visit_date: Timestamp) -> Timestamp {
+    let now = Timestamp::now();
+    if visit_date > now {
+        return visit_date;
+    }
+    if visit_date < EARLIEST_TIMESTAMP {
+        return EARLIEST_TIMESTAMP;
+    }
+    return visit_date;
+}
+
+#[derive(Debug)]
+pub enum IncomingPlan {
+    Skip, // An entry we just want to ignore - either due to the URL etc, or because no changes.
+    Invalid(Error), // Something's wrong with this entry.
+    Failed(Error), // The entry appears sane, but there was some error.
+    Delete(), // We should delete this.
+    // We should apply this. If SyncGuid is Some, then it is the existing guid
+    // for the same URL, and if that doesn't match the incoming record, we
+    // should change the existing item to the incoming one and upload a tombstone.
+    Apply(Option<SyncGuid>, Url, Option<String>, Vec<HistoryRecordVisit>),
+    Reconciled, // Entry exists locally and it's the same as the incoming record.
+}
+
+
+fn plan_incoming_record(conn: &Connection, record: HistoryRecord, max_visits: usize) -> IncomingPlan {
+    let url = match Url::parse(&record.hist_uri) {
+        Ok(u) => u,
+        Err(e) => return IncomingPlan::Invalid(e.into()),
+    };
+    match can_add_url(&url) {
+        Ok(can) => if !can { return IncomingPlan::Skip },
+        Err(e) => return IncomingPlan::Failed(e.into()),
+    }
+    // Let's get what we know about it, if anything - last 20, like desktop?
+    let visit_tuple = match fetch_visits(conn, &url, max_visits) {
+        Ok(v) => v,
+        Err(e) => return IncomingPlan::Failed(e.into()),
+    };
+
+    // This all seems more messy than it should be - struggling to find the
+    // correct signature for fetch_visits
+    let (existing_page, existing_visits): (Option<FetchedVisitPage>, Vec<FetchedVisit>) = match visit_tuple {
+        None => (None, Vec::new()),
+        Some((p, v)) => (Some(p), v),
+    };
+
+    let mut old_guid: Option<SyncGuid> = None;
+    let mut cur_visit_map: HashSet<(VisitTransition, Timestamp)> = HashSet::new();
```

Comment:
We do have an expected capacity for this, which is `existing_visits.len()`, so it's probably worth doing `HashSet::with_capacity(existing_visits.len())`